Pass trailing `x run --` arguments to launch command

MarijnS95 opened the pull request:
If you want to pass string "extras" into an Android intent via `am start`, call `x run ... -- -e extraKey "extra string"` (and similar for other types like `--ez key bool` for booleans, or `-d URI` for data URIs).  [All arguments listed by `am help` under the `start` command](https://developer.android.com/tools/adb#am) (including the `INTENT` specification) are supported.

On the host, this passes the arguments as-is to the running executable, similar to `cargo run .. -- <launch_args>`.

For iOS, these are passed to [`idevicedebug run`](https://manpages.debian.org/testing/libimobiledevice-utils/idevicedebug.1.en.html).

MarijnS95 committed:
Pass trailing `x run --` arguments to launch command

If you want to pass string "extras" into an Android intent via `am
start`, call `x run ... -- -e extraKey "extra string"` (and similar
for other types like `--ez key bool` for booleans, or `-d URI` for data
URIs).  [All arguments listed by `am help` under the `start` command]
(including the `INTENT` specification) are supported.

On the host, this passes the arguments as-is to the running executable,
similar to `cargo run .. -- <launch_args>`.

For iOS, these are passed to [`idevicedebug run`].

[All arguments listed by `am help` under the `start` command]: https://developer.android.com/tools/adb#am
[`idevicedebug run`]: https://manpages.debian.org/testing/libimobiledevice-utils/idevicedebug.1.en.html

MarijnS95 committed:
Fix simple clippy lints

MarijnS95 commented on xbuild/src/devices/adb.rs:
```diff
@@ -129,16 +129,25 @@ impl Adb {
             .arg("-n")
             .arg(format!("{}/{}", package, activity))
-            .status()?;
+            // Quote arguments for `am` so that they don't already get parsed by `adb`.
+            // TODO: use shlex quoting helpers? https://docs.rs/shlex/1.3.0/shlex/
+            .args(launch_params.iter().map(|s| format!("\"{s}\""))))
```
This must be fixed. I quickly tested some hacky command strings and it already falls apart.

MarijnS95 replied:
For example, we have `-- -e cmdLine "some-subcommand --foo"`. If I insert the quotes such that the `--help` command becomes "free-standing" with `-- -e cmdLine "some-subcommand\" --foo\""` or similar, because my wrapping in code here "inverts" the quoted selection (end-result is `"subcommand" --foo""`), `adb` splits our command in two and `am start` doesn't know what `--foo` is. `shlex` should be able to re-quote and/or escape this string such that it always gets seen as `"subcommand --foo"` by `am start`.

MarijnS95 replied:
Fixed with `shlex`. Note that the above borked example at least prevents `adb` or `am` from parsing our string, and instead gives the `some-subcommand\" --foo\"` to our `Activity` as expected (to be read via https://github.com/rust-mobile/android-intent/pull/4). `shlex` on the reading side will "split" that as a single `["some-subcommand --foo"]` argument which obviously isn't valid. That's why it shouldn't have been double-quoted :) (but again: was a good way to test if we still clobber `adb` or `am` argument quoting here).

Jasper-Bekkers reviewed (approved):
Approving, but could you document in the PR description (and the `x` help`) exactly how to use this feature?

MarijnS95 commented:
@Jasper-Bekkers this was already somewhat explained in the PR description but the `--help` text now elaborates where these arguments end up at runtime, **per platform**, and links to the platform-specific documentation.

MarijnS95 committed:
CI: Bump `upload-artifact` to v4

Starting today (January 30th, 2025), v3 is no longer available to use
after a long deprecation period:
https://github.blog/changelog/2024-04-16-deprecation-notice-v3-of-the-artifact-actions/
